File: RankingGenerator.py
from TournamentDescriptionClasses import Game
import random
import sys
import matplotlib.pyplot as plt
from typing import List

# conversion table: difference in rank to expected result of matchup
rankDeltaToResultDelta = [
    0,
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
    13,
    13,
    13,
    13,
    13,
    13,
]


def _swap(list: List[str], indexA: int, indexB: int):
    """ swap two list members
    """
    temp = list[indexA]
    list[indexA] = list[indexB]
    list[indexB] = temp


def _calculateTotalError(currentRanking: List[str], games: List[Game]) -> float:
    """ Error between obtained results and results expected from ranking
    Takes into account all obtained results so far
    Converts given Ranking into results using conversion table
    Calculates quadratic error
    """
    lossSum = 0
    for game in games:
        rankA = currentRanking.index(game.matchup.first)
        rankB = currentRanking.index(game.matchup.second)
        rankDelta = rankA - rankB
        resultDelta = game.result.first - game.result.second
        predictedResultDelta = rankDeltaToResultDelta[abs(rankDelta)]
        if rankDelta > 0:
            predictedResultDelta *= -1
        lossSum += (resultDelta - predictedResultDelta)**2
    return lossSum


def generateNewRankingPython(currentRanking: List[str], games: List[Game], debug=False) -> List[str]:
    """ heuristic search for an optimal Ranking to explain the results
    compares random permutations in their total error
    uses some variant of simulated annealing for the search
    XXX possible time gain: better determination criteria, Tabu-List
    XXX pAcceptWorse=4.5e-5 after 1e5 iterations, currently 1e6 iterations are set
    """
    # list of errors made by different rankings
    losses = []
    # current elements in the search
    minimalLoss = sys.maxsize
    minimalLossRanking = currentRanking[:]
    newRanking = currentRanking[:]
    # tuning parameters for search
    pAcceptWorse = 0.5 # probability to accept a worse result
    decay = 0.9999     # decay parameter for pAcceptWorse
    if debug:
        print("Original ranking:")
        for item in currentRanking:
            print(item)
        print("initial Erorr: loss=", _calculateTotalError(currentRanking, games))
    # cover large range of possible random permutations:
    for i in range(0,150000):
        # calculate total Error made by current ranking
        currentLoss = _calculateTotalError(newRanking, games)
        losses.append(currentLoss)
        # randomly permute the ranking
        indexA = random.randint(0, len(currentRanking)-1)
        indexB = indexA
        while indexA == indexB:
            indexB = random.randint(0, len(currentRanking)-1)
        _swap(newRanking, indexA, indexB)
        # calculate total error of new ranking
        newLoss = _calculateTotalError(newRanking, games)
        # if better: save optimal solution
        if newLoss < minimalLoss: 
            minimalLoss = newLoss
            minimalLossRanking = newRanking[:]
        # possibly use worse ranking (to avoid local minima)
        if newLoss > currentLoss and random.uniform(0.0, 1.0) > pAcceptWorse: # do not accept worse value
            _swap(newRanking, indexA, indexB)
        pAcceptWorse *= decay
    if debug: # debug messaging
        print("Optimal Ranking:")
        for item in minimalLossRanking:
            print(item)
        print("Optimal Error: loss=", minimalLoss)
        print("pAcceptWorse", pAcceptWorse)
        fig1,ax1=plt.subplots()
        ax1.plot(losses)
        ax1.set_xlabel('iteration #')
        ax1.set_ylabel('total error')
        ax1.set_title('Error for different permutations')
        plt.show()
    return minimalLossRanking

def generateNewRankingCpp(currentRanking: List[str], games: List[Game], debug=False) -> List[str]:
    teamNameToInt = dict()
    for number, teamName in enumerate(currentRanking, 0):
        teamNameToInt[teamName] = number

    currentRankingInt = list()
    for teamName in currentRanking:
        currentRankingInt.append(teamNameToInt[teamName])

    gamesInt = list()
    for game in games:
        gameInt = RankingGeneratorCpp.GameInt(teamNameToInt[game.matchup.first],
                                    teamNameToInt[game.matchup.second],
                                    game.result.first,
                                    game.result.second)
        gamesInt.append(gameInt)

    rankingGeneratorCpp = RankingGeneratorCpp.RankingGenerator()
    rankingInt = rankingGeneratorCpp.generateRanking(currentRankingInt, gamesInt, debug)

    newRanking = list()
    for number in rankingInt:
        for key, value in teamNameToInt:
            if value == number:
                newRanking.append(key)
                break
    return newRanking

try:
    import RankingGeneratorCpp
    rankingFunction = generateNewRankingCpp
except ImportError:
    rankingFunction = generateNewRankingPython

def generateNewRanking(currentRanking: List[str], games: List[Game], debug=False) -> List[str]:
    return rankingFunction(currentRanking, games, debug)

File: cpp/RankingGenerator.cpp
#include "RankingGenerator.h"
#include <algorithm>
#include <cmath>
#include <iostream>
#include <random>

std::vector<TeamInt>
RankingGenerator::generateRanking(const std::vector<TeamInt> &currentRanking,
                        const std::vector<GameInt> &games,
                        const bool debug) {
  std::mt19937 gen(18); // use constant for init so ranking does not jump around, although input was the same as before
  std::uniform_int_distribution<> distribution(0, currentRanking.size() - 1);
  std::uniform_real_distribution<> acceptDist(0.0, 1.0);
  std::vector<double> losses;
  // current elements in the search
  double minimalLoss = std::numeric_limits<double>::max();
  std::vector<TeamInt> minimalLossRanking = currentRanking;
  std::vector<TeamInt> newRanking = currentRanking;
  // tuning parameters for search
  double pAcceptWorse = 0.5; // probability to accept a worse result
  double decay = 0.9999;     // decay parameter for pAcceptWorse
  if(debug) {
    //print("Original ranking:")
    for(const auto &item : currentRanking) {
      //print(item)
    }
    std::cout << "initial Erorr: loss=" << calculateTotalError(currentRanking, games) << std::endl;
  }
  // cover large range of possible random permutations:
  for(int i = 0; i < 150000; ++i) {
    // calculate total Error made by current ranking
    const auto currentLoss = calculateTotalError(newRanking, games);
    losses.push_back(currentLoss);
    // randomly permute the ranking
    const int indexA = distribution(gen);
    int indexB = indexA;
    while(indexA == indexB) {
      indexB = distribution(gen);
    }
    std::iter_swap(newRanking.begin() + indexA, newRanking.begin() + indexB);
    // calculate total error of new ranking
    const double newLoss = calculateTotalError(newRanking, games);
    // if better: save optimal solution
    if(newLoss < minimalLoss) {
      minimalLoss = newLoss;
      minimalLossRanking = newRanking;
    }
    // possibly use worse ranking (to avoid local minima)
    if(newLoss > currentLoss && acceptDist(gen) > pAcceptWorse) { // do not accept worse value
      std::iter_swap(newRanking.begin() + indexA, newRanking.begin() + indexB);
    }
    pAcceptWorse *= decay;
  }

  if(debug) // debug messaging
  {
      std::cout << "Optimal Ranking:" << std::endl;
      for(const auto& item : minimalLossRanking)
          std::cout << item.team << std::endl;
      std::cout << "Optimal Error: loss=" <<  minimalLoss << std::endl;
      std::cout << "pAcceptWorse: " << pAcceptWorse << std::endl;
  }
  //TODO write results to plotable file (eg. for gnuplot)
/*
      fig1,ax1=plt.subplots()
      ax1.plot(losses)
      ax1.set_xlabel('iteration #')
      ax1.set_ylabel('total error')
      ax1.set_title('Error for different permutations')
      plt.show()
  */
  for(int i = 0; i < minimalLossRanking.size(); ++i) {
    minimalLossRanking[i].newRank = i+1;
  }
  return minimalLossRanking;
}

double RankingGenerator::calculateTotalError(const std::vector<TeamInt> &currentRanking, const std::vector<GameInt> &games) {
  double lossSum = 0;

  for(const auto& game : games) {
    //TODO build lookup table for team to rank conversion
    const auto rankAIt = std::find_if(currentRanking.begin(), currentRanking.end(),
            [&](const auto& value) { return value.team == game.teamA;});
    const auto rankBIt = std::find_if(currentRanking.begin(), currentRanking.end(),
            [&](const auto& value) { return value.team == game.teamB;});
    const int rankA = std::distance(currentRanking.begin(), rankAIt);
    const int rankB = std::distance(currentRanking.begin(), rankBIt);

    const int rankDelta = rankA - rankB;
    const int resultDelta = game.resultA - game.resultB;
    double predictedResultDelta = m_rankDeltaToResultDelta[std::abs(rankDelta)];
    if(rankDelta > 0) {
      predictedResultDelta *= -1;
    }
    lossSum += std::pow(resultDelta - predictedResultDelta, 2);
  }

  const int approximatedSwissRound = std::ceil((double)games.size() / (double)currentRanking.size());
  // ranking in seed is expected to start at 1
  for(int rank = 0; rank < currentRanking.size(); ++rank) {
    if(approximatedSwissRound <= 1) {
      lossSum += pow((rank+1) - currentRanking[rank].seed, 2);
    }
    else if(approximatedSwissRound == 2){
      lossSum += pow((rank+1) - currentRanking[rank].seed, 2) / 4.0;
    }
  }
  return lossSum;
}
